modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java

```diff
@@ -164,4 +248,16 @@ private boolean isReplicationSetAlreadyInitialized() {
         );
         return execCheckRsInit.getExitCode() == CONTAINER_EXIT_CODE_OK;
     }
+
+    @Builder
+    @Getter
+    public static class ConnectionString {
+
+        @Builder.Default
+        private final String databaseName = DEFAULT_DATABASE_NAME;
+
+        private final String username;
+
+        private final String password;
+    }
```

Comment:
I'm not sure that it's a good idea here because:

- Exposing `MongoClientSettings` to the public API seems to be dangerous because a user might set a SRV DNS or other settings that `MongoDBContainer` does not support. I'd rather not allow a user to construct a URL on their side, it's safer to do delegate it to `MongoDBContainer` providing a URL string. Otherwise, we might receive some unexpected bug tickets here;

- `ConnectionString` is for the public API purpose to expose only `databaseName`, `username` and `password` for now. As apposed to `MongoClientSettings`, it doesn't require setting auth db source, database etc. which simplifies the public API and takes control of it

Reply:
> I'd rather not allow a user to construct a URL on their side, it's safer to do delegate it to MongoDBContainer providing a URL string. Otherwise, we might receive some unexpected bug tickets here;

TBH, I do not expect users to do that but who knows.

I think our docs can be improved in order to make it clear

Reply:
We provide a URL string for a user via clear API that we take control of. I reckon it's not a good idea to constantly update docs mentioning "it does not support srv etc."? In my opionin, it's better not to provide a way to set srv.

> I was not proposing to use Mongo API in MongoDBContainer. Just sharing how it can be done from the outside as in [this suggestion](https://github.com/testcontainers/testcontainers-java/pull/5595#discussion_r1049049744)

I see your point, but it's not a part of our public API. This way a user can construct everything, but having an issue with it is on their hands

Reply:
> > I'd rather not allow a user to construct a URL on their side, it's safer to do delegate it to MongoDBContainer providing a URL string. Otherwise, we might receive some unexpected bug tickets here;
> 
> TBH, I do not expect users to do that but who knows.
> 
> I think our docs can be improved in order to make it clear

I updated Java doc to make it clear

Reply:
Sorry @silaev, but it is still unclear with me what is the issue with the approach suggested by @eddumelendez, since it is very close to how we do it for other modules.

So you think users will struggle doing it this way and construct invalid URLs?

Reply:
Hi @kiview The idea of this builder is to provide 3 params for now (might be more in the future): username, pass and db name for MongoDBContainer to generate a URL as simple as this:
```
MongoClients.create(
    mongoDBContainer.getReplicaSetUrl(
        MongoDBContainer.ConnectionString
            .builder()
            .username(usernameRestrictedAccess)
            .password(passwordRestrictedAccess)
            .build()
    )
)
```
It’s similar to withUsername(...), withPassword(…) and withDatabaseName(...) etc. builder methods in the PostgreSQLContainer to generate a URL. As I understand, generating URL inside a container is `close to how we do it for other modules`. There is already ```public String getReplicaSetUrl(final String databaseName) {}```, so I followed the same approach with a param here ```public String getReplicaSetUrl(final ConnectionString connectionString)```. 
  
@eddumelendez proposed a solution when a URL is generated outside of MongoDBContainer via MongoClientSettings which we do not control. Is it `close to how we do it for other modules` ? This way a user has to construct a URL on their own and know some additional data:
```
MongoClients.create(
    MongoClientSettings.builder()
        .credential(MongoCredential.createCredential(usernameRestrictedAccess, "admin", passwordRestrictedAccess.toCharArray()))
        .applyConnectionString(new ConnectionString(String.format("mongodb://%s:%d", mongoDBContainer.getHost(), mongoDBContainer.getFirstMappedPort()) ))
        .build()
)
```

The second variant seems to me as quite complex and error prone, so I'm more comfortable with the first one. Correct me if I'm wrong, but the whole idea of Testcontainers is to simplify testing experience which, as I see it, the first variant follows

Reply:
Hi @kiview , @eddumelendez
I haven’t heard back from you after [my last comment](https://github.com/testcontainers/testcontainers-java/pull/5595#discussion_r1051061701). Did I have a change to convince you to go with the simple builder approach?

Reply:
Hi, thanks for ping us, I was thinking about it last week. There is a difference with the builders in database modules due to those are used by `JdbcDatabaseContainer` for the specific wait strategy and, well, JDBC API is in the JDK itself.

With latest changes in the `MongoDbContainer`, I would suggest the following:
- Make authentication opt-in, `withAuthenticationEnable()`, instead of making it the default. Unless, a specific MongoDB version is forcing it, we can check on that version.
- Provide `withUsername` and `withPassword` as initially provided in the PR
- Do not provide a default `key` file. So, the user is aware of what's involved to run MongoDB with authentication enabled.
- Make it work with sharding, if it is possible
- Improve our javadocs and docs in general to show the correct usage of the module.

Testcontainers responsibility is to provide a ready to use service running in a container. There could be many configuration in the service itself which we do not master and we prefer a generic approach, meanwhile we are learning from uses cases, rather than do something at the beginning that can change later just because we are not aware of it.

I know the answer doesn't answer the last question but I think it is our best effort, meanwhile we also learn about MongoDB and how to use it. So, the builder can be revisited in the future. I'm also aware that this will involve more changes than initially expected but it is due to the evolution of the module itself.